### DISCUSSION TITLE
After upgrading from webpack v4 to v5, getting "Uncaught TypeError: Cannot read properties of undefined (...)"

### DISCUSSION THREAD

**Discussion by russ0133** [webpack/webpack#16717] [Q&A] [answered]
# Bug report

**What is the current behavior?**
Upgrading from webpack v4 to v5 through upgrading react-scripts from v4 to v5 renders only a blank page while giving the following errors:

```
index.tsx:7 Uncaught TypeError: Cannot read properties of undefined (reading 'ApiKeyRow')
    at Module.ApiKeyRow (index.tsx:7:1)
    at Module.ApiKeyRow (index.tsx:78:1)
    at registerExportsForReactRefresh (RefreshUtils.js:146:1)
    at Object.executeRuntime (RefreshUtils.js:177:1)
    at $ReactRefreshModuleRuntime$ (index.ts:2:1)
    at ./src/shared/index.ts (index.ts:2:1)
    at options.factory (react refresh:6:1)
    at __webpack_require__ (bootstrap:24:1)
    at fn (hot module replacement:62:1)
    at ./src/shared/ui/Card/EndpointCard/index.jsx (index.tsx:17:1)
```

Mind you, this is not specific to any react components. If I delete ApiKeyRow component, for example, it will just choose another random component to complain about.

It's not something with the component itself, since the error persists even after removing all code and returning only a <div>test</div>. I have upgraded all my plugins.

**If the current behavior is a bug, please provide the steps to reproduce.**
What I did is run npm install react-scripts@5.0.0, which also updates Webpack to v5 from v4. I have almost confirmed this isn't a react-scripts problem (by ejecting and trying to upgrade webpack myself), but a webpack problem.

**What is the expected behavior?**
It should render the website normally.

**Other relevant information:**
webpack version:  5.64.4 (used in react-scripts v5.0.0)
Node.js version: 16.19.0
Operating System: Ubuntu
Additional tools: create-react-app (react-scripts)
Package.json:
   ```
 "@emotion/react": "^11.10.5",
    "@emotion/styled": "^11.10.5",
    "@feedback-fish/react": "^1.2.1",
    "@mui/icons-material": "^5.4.2",
    "@mui/lab": "^5.0.0-alpha.79",
    "@mui/material": "^5.6.3",
    "@mui/styles": "5.4.1",
    "@mui/system": "^5.10.3",
    "@mui/x-data-grid-pro": "^5.9.0",
    "@mui/x-date-pickers-pro": "^5.0.0-beta.3",
    "@mui/x-license-pro": "^5.7.0",
    "@reduxjs/toolkit": "^1.6.2",
    "@sentry/react": "^7.12.1",
    "@sentry/tracing": "^7.12.1",
    "@testing-library/jest-dom": "^5.11.4",
    "@testing-library/react": "^11.1.0",
    "@testing-library/user-event": "^12.1.10",
    "@types/jest": "^26.0.15",
    "@types/js-cookie": "^3.0.2",
    "@types/react-redux": "^7.1.20",
    "antd": "^4.19.2",
    "aos": "^3.0.0-beta.6",
    "axios": "^0.26.0",
    "classnames": "^2.3.1",
    "compose-function": "^3.0.3",
    "cross-env": "^7.0.3",
    "date-fns": "^2.27.0",
    "effector": "^22.1.2",
    "effector-react": "^22.0.5",
    "js-cookie": "^3.0.1",
    "js-file-download": "^0.4.12",
    "lodash.clonedeep": "^4.5.0",
    "notistack": "^2.0.3",
    "react": "^17.0.2",
    "react-calendly": "^4.1.0",
    "react-dom": "^17.0.2",
    "react-google-button": "^0.7.2",
    "react-google-login": "^5.2.2",
    "react-helmet": "^6.1.0",
    "react-intersection-observer": "^8.33.1",
    "react-js-cron": "^1.3.1",
    "react-json-view": "^1.21.3",
    "react-redux": "^7.2.6",
    "react-router": "^6.0.2",
    "react-router-dom": "^6.0.2",
    "react-scripts": "4.0.3",
    "react-simple-oauth2-login": "^0.5.3",
    "react-syntax-highlighter": "^15.5.0",
    "typescript": "^4.1.2",
    "uuid": "^8.3.2"
```

**Comment by alexander-akait** [comment]
Sorry, without reproducble test repo we can't help. Did you try to debug the line in the error?

**Reply by russ0133** [reply]
@alexander-akait first of all, thanks for your help. As I mentioned, the line is just a react component that returns a div with "Test" (I changed it to debug). If I delete it, it accuses the same error in a different component.

Some guys told me this might be a circle import problem, but I'm not sure yet. Thank you again!

**Reply by russ0133** [reply]
Sadly, I can not give the full code because that's company code. I'm not expecting an exact answer either, just maybe if you guys from webpack can point me to the right direction on WHY the problem is happening, it would be great.

**Reply by alexander-akait** [reply]
Do you got such problems for production or development build?

**Reply by russ0133** [reply]
> Do you got such problems for production or development build?

I am in development just trying to run the app with npm run start.

**Reply by alexander-akait** [reply]
I think better create an issue in `create-react-app`, without reproducible example I can't help, because there are a lot of different things can happen, as minimum can you provide a code there it happens?

**Reply by laustdeleuran** [reply]
I might be running into the same issue, and I have a reproducible test repo. More info in https://github.com/webpack/webpack/issues/16776.

**Reply by russ0133** [reply, marked as answer]
I've found a solution.

it ends up that in my repository, we have a couple stray .jsx files (it's a typescript project), and our folder structure is in such a way:

shared
- index.ts (exporting everything below)
-- lots_of_component_folders
----> index.tsx

And the stray .jsx files were trying to import some compoments exported from shared/index.ts and that wasn't working, so after I pointed to the folder directly instead of to the /shared/ folder, it worked.

For example:
import { Component } from "shared"; <<<< BUG
import { Component } from "shared/ui/Component"; <<<< WORKS

Disabling FAST_REFRESH by creating a .env and putting FAST_REFRESH=false also worked, but I didn't want to have this disabled so I ended up fidnign this solution.
